AAD Authentication using AccessToken

paulmey opened the pull request:
Although it leaves room for improvement with respect to convenience from the DSN, this PR implements AAD authentication through access tokens.
To test this implementation, create a SQL Azure database and set yourself as the AAD admin for the database. Then run
``` bash
az login # if not already logged in
at=$(az account get-access-token --resource https://database.windows.net/ --q accessToken -o tsv)
export SQLSERVER_DSN="sqlserver://srv.database.windows.net?database=db&accesstoken=$at"
go test -v
```
To use a service principal, grant the service principal access to the database as described in the [docs](https://docs.microsoft.com/en-us/azure/sql-database/sql-database-aad-authentication-configure?tabs=azure-powershell#create-contained-database-users-in-your-database-mapped-to-azure-ad-identities).

At least partially fixes #446

paulmey commented:
I'll try to raise code coverage a bit

kardianos commented on token.go:
```diff
@@ -447,6 +448,18 @@ func parseLoginAck(r *tdsBuffer) loginAckStruct {
 }
 
+func parseFeatureExtAck(r *tdsBuffer) {
+	for {
```
I don't like unbounded loops when parsing inputs.

Can we put some ridiculous large constraint here like 1000 or even just 10?
for i := 0; i < 10; i++ {
   ...
}
panic("parsed more then X and we are worried!")

paulmey replied:
I think that's fair. I'll go back to the spec and see if there is a maximum number of features that is accepted, most likely each possible feature can only be present once in the login packet, so only once in the login response.

kardianos commented on token.go:
```diff
@@ -447,6 +448,18 @@ func parseLoginAck(r *tdsBuffer) loginAckStruct {
+	for {
+		featureID := r.byte() // FeatureID
+		if featureID == 0xff {
+			break
```
Just "return" please.

paulmey commented on README.md:
```diff
@@ -106,6 +106,26 @@ Other supported formats are listed below.
+
+The configuration of functionality might change in the future.
+
+Azure Active Directory (AAD) is supported through the `accesstoken` parameter in the DSN. However,
```
I had the DSN hard coding before the connector. I like the connector better as well and in #548 I actually removed it. It was useful for testing, however, to be able to run all the integration tests with an access token, but I think that #547 takes care of that.
If we want to merge this PR before #547, I'll remove the DSN parameter and let the struct names align with #547.

paulmey commented on tds.go:
```diff
@@ -844,15 +958,22 @@ initiate_connection:
+	auth, authOk := getAuth(p.user, p.password, p.serverSPN, p.workstation)
+	if p.accessToken != "" { // accesstoken ignores user/password
+		featurext := &featureExtFedAuthSTS{
+			FedAuthEcho:  fields[preloginFEDAUTHREQUIRED] != nil && fields[preloginFEDAUTHREQUIRED][0] == 1,
```
Good point, i'll add the check. IMHO either SQL or the open spec is broken if that were true, but I agree it would be sad to panic...

paulmey commented:
I added a go.mod in the specific example folder and fixed most of the nits. The go.mod in the example need to be removed to make the example work until this gets merged, but hopefully that's only going to be temporary.
